tao_compiler/mlir/disc/IR/disc_shape_ops.cc

```diff
@@ -296,6 +298,68 @@ void TieShapeOp::getCanonicalizationPatterns(RewritePatternSet& results,
 
 LogicalResult TieShapeOp::verify() { return Verify(*this); }
 
+//===----------------------------------------------------------------------===//
+// SymbolicDimOp
+//===----------------------------------------------------------------------===//
+
+void SymbolicDimOp::getCanonicalizationPatterns(RewritePatternSet& results,
+                                                MLIRContext* context) {}
+
+LogicalResult SymbolicDimOp::verify() { return Verify(*this); }
+
+int64_t SymbolicDimOp::getDimSize() {
+  if (auto attr = (*this)->getAttrOfType<IntegerAttr>("value"))
+    return attr.getInt();
+  return ShapedType::kDynamicSize;
+}
+
+void SymbolicDimOp::setDimSize(int64_t val) {
+  OpBuilder b(*this);
+  (*this)->setAttr("value", b.getI64IntegerAttr(val));
+  if (val == -1) {
+    setKnownNegativeOne(true);
+  }
+}
+
+bool SymbolicDimOp::isDynamic() {
+  return getDimSize() == ShapedType::kDynamicSize;
+}
+
+void SymbolicDimOp::setKnownNonNegative(bool flag) {
+  OpBuilder b(*this);
+  (*this)->setAttr("knownNonNegative", b.getBoolAttr(flag));
+}
+
+void SymbolicDimOp::setKnownNegativeOne(bool flag) {
+  OpBuilder b(*this);
+  (*this)->setAttr("knownNegativeOne", b.getBoolAttr(flag));
+  if (flag) (*this)->setAttr("knownNonSizeOne", b.getBoolAttr(flag));
```

Comment:
Just curious, what is the relationship between "knowNegativeOne" and "knownNonSizeOne"? 
It seems not very intuitive for me to comprehend.

Reply:
Background:
For op like `reshape`, frontends (e.g. TF or PyTorch) support at most one `-1` in the target shape. For example, `reshape(tensor<10x10>, [2, 25, -1])` is equal to `reshape(tensor<10x10>, [2, 25, 2])`. In dynamic shape semantics, we have to take care of such `-1` when lowering `reshape` like ops. If we can know all dims in the target shape are not `-1`, then we could simplify the lowering of reshape.

For op like `broadcast_in_dim`, `1` has special meaning. For example, `[1, 10] +[10, 10]`  is valid, while `[2, 10] + [10, 10]` is not valid. For a pattern`[a, b] + [c, d] = [e, f]`, if we known `a` is not `1`, then we can know that `e == c`.